shared/src/store/action-reducers/exercise.ts

```diff
@@ -77,19 +78,38 @@ export namespace ExerciseActionReducers {
         ) => {
             // Refresh the current time
             draftState.currentTime += tickInterval;
+
             // Refresh patient status
             patientUpdates.forEach((patientUpdate) => {
                 const currentPatient = draftState.patients[patientUpdate.id]!;
+
+                const visibleStatusBefore = Patient.getVisibleStatus(
+                    currentPatient,
+                    draftState.configuration.pretriageEnabled,
+                    draftState.configuration.bluePatientsEnabled
+                );
+
                 currentPatient.currentHealthStateId = patientUpdate.nextStateId;
                 currentPatient.health = patientUpdate.nextHealthPoints;
                 currentPatient.stateTime = patientUpdate.nextStateTime;
                 currentPatient.treatmentTime = patientUpdate.treatmentTime;
                 currentPatient.realStatus = getStatus(currentPatient.health);
+
+                const visibleStatusAfter = Patient.getVisibleStatus(
+                    currentPatient,
+                    draftState.configuration.pretriageEnabled,
+                    draftState.configuration.bluePatientsEnabled
+                );
+
+                if (
+                    refreshTreatments &&
+                    // We only want to do this expensive calculation, when it is really necessary
+                    visibleStatusBefore !== visibleStatusAfter
+                ) {
+                    updateTreatments(draftState, currentPatient);
+                }
```

Comment:
As @anonym-HPI rightly noticed, my changes here were incorrect. Currently, this doesn't take into account changes in the `visualStatus` that happened when `refreshTreatments` is `false` (which happens in 19 out of 20 cases).

What about removing `refreshTreatments` altogether and always updating the treatments directly after they happen? While this would be a bit less realistic (personnel would only seldom change the patient they are treating instantaneously), the same would also apply to changing the triage status of a patient or moving a patient.
The current behavior also encourages micro-management (very low-level decisions, which shouldn't be done anymore once the simulation is a few minutes in).
Removing `refreshTreatments` would make it much less complex, the code more maintable and the behavior more predictable.
Because `updateTreatments` is now (thanks to @anonym-HPI ) much more efficient, and status changes are happening only rarely (`< numberOfPatients * 3` per exercise could be a good approximation), this should still be as fast as it is now.

What are your thoughts about it @anonym-HPI ?

Reply:
Yeah, I forgot the "treatment-jumping-problem".
I think this would be prevented by #490 because a patient will (for most configurations) stay at his category for a certain amount of time until it can change.
I'm still unsure about this...

About the performance: 
The implementation with `refreshTreatments` is (with your `visibleStatus` check improvement) only more performant if `visibleStatus` changes more than once in 20 seconds for a patient. Correct? I would assume that this is only seldom the case.

Reply:
No, the `visibleStatus` improvement means a patient gets a treatment update only max 4 times in a complete exercise (in a tick) instead of every time `refreshTreatment` is true.

Of course max 4 times if patients only get worse (e.g. pretriaged as yellow, are treated swapping to true category green, getting worse to yellow and then to red and dying afterwards).


Of couse overhead is created for checking the `visibleStatus` before and after and comparing.

Reply:
> No, the `visibleStatus` improvement means a patient gets a treatment update only max 4 times in a complete exercise (in a tick) instead of every time `refreshTreatment` is true.
> 
> Of course max 4 times if patients only get worse (e.g. pretriaged as yellow, are treated swapping to true category green, getting worse to yellow and then to red and dying afterwards).
> 
> Of couse overhead is created for checking the `visibleStatus` before and after and comparing.

I didn't mean the comparison "with `visibleStatus` improvement vs without `visibleStatus` improvement" but instead "`visibleStatus` improvement with `refreshTreatments` vs `visibleStatus` improvement without `refreshTreatments`".

Reply:
Yes, thats true.

What do you say we use something like id of patient (as number) + current exercise time modulo 20?

This would remove refreshTreatment but would solve the jumping problem (they would only jump every 20 seconds).

But this could be (more) confusing as careringLines could gradually switch.

Reply:
> What do you say we use something like id of patient (as number) + current exercise time modulo 20?

To remove `refreshTreatments` we could also just do `exerciseTime % 20 === 0` - this would move the calculation `refreshTreatments` from the backend to the reducer (which is better IMO).

My main objection was that the additional variable for each patient would not be worth the benefit (more code, more complexity vs. a minimal performance improvement, and a fix for a shortcoming of the health point system). In addition, we would also have to write a migration for it.

I'd let you decide which option we choose:
1. Each patient has an additional `visibleStatusChangedSinceTreatment` variable 
2. We update the changed treatments for every tick.

Reply:
They are 4 possibilities with `refreshTreatment` and `visibleStatusChangedSinceTreatment`:
1. using both
2. using only `refreshTreatment`
3. using only `visibleStatusChangedSinceTreatment`
4. using none of the two

Lets look at numbers of updateTreatments called via `tick`.
First some numbers:
-  a.) Exercise time: 2 hours (7200 sec. = 7200 ticks) 
-  b.) Number of patients: 20
-  c.) only patients, patients not moving
-  d.) short: every patient has 4 category changes in the exercise
- e.) `refreshTreatments` is `true` every 20th tick (like right now)
- f.) a `tick` is called every second (like right now)

Description for d.):
- all patients start as green and get worse (and only worse) roughly every 30 min.
- after 2 hours every patient dies (lets assume all die at the last tick)
- this gives every patient green to yellow, yellow to red, red to black - 3 changes
- lets just add one for pretriage, which would result in 4 changes per patient in the exercise

Number of `updateTreatments` (before `calculateTreatments` called:
1. 4 * 20 = 80 (for every category change will `updateTreamtents` be called)
2. 7200 * 20 / 20 = 7200 (every 20th tick will for every of the 20 patients `updateTreatments` be called)
3. 4 * 20 = 80 (for every category change will `updateTreamtents` be called)
4. 7200 * 20 = 144.000 (every tick for every patient `updateTreatments` is called)

As you can see, there is a big difference in the number of `updateTreatments` calls when `visibleStatusChangedSinceTreatment` is not used.

Of course using `visibleStatusChangedSinceTreatment` creates the overhead of calculating `getVisibleStatus` for every patient in every `tick` so 288.000 times and every 20th tick the if statement has to compare one `boolean` to `boolean` more - 7.200 times (when `refreshTreatments` is also used, else it would be 144.000).

`1` and `3` seem to be the same, only difference is that (apart from the earlier stopped if statement, because `refreshTreatments` is written before) `1` would make jumping happen only for every 20 seconds and `3` every second.
This of course only, when jumping accours (when @hpistudent72 thesis is merged this should mostly not be the case, but I am not sure his changes forbid such a behaviour in the future).
`1` and `3` would only differ in the number of `updateTreatments` when a patient changes the category more than once in 20 seconds.
Or when a patient was moved while `visibleStatusChangedSinceTreatment` is true. In this case `visibleStatusChangedSinceTreatment` would be set to false, as moving a patient updated him already.

We can put the calculation of `refreshTreatments` into the reducer, yes.

Which version do you @Dassderdie prefer (maybe it changed after this example?)?

Reply:
I decided to keep `refreshTreatments` for now -> 1 it is.
See c960ecc5b90b9bf12619c6c2c76d839044041ee3